File: csharp/DataStructures/src/dsa/matrix/SpiralMatrix.cs
namespace dsa.matrix;

    public class SpiralMatrix
    {
        // static void Main()
        // {
        //     int[][] matrix = {
        //         new int[] { 1, 2, 3, 4 },
        //         new int[] { 5, 6, 7, 8 },
        //         new int[] { 9, 10, 11, 12 },
        //         new int[] { 13, 14, 15, 16 }
        //     };
        //     spiralPrint(matrix, matrix.Length, matrix[0].Length);
        // }

        public static void spiralPrint(int[][] matrix, int r, int c)
        {
            int i;
            int k = 0; // k -> r
            int l = 0; // l -> c

            // Enquanto houver linhas e colunas para serem percorridas
            while (k < r && l < c)
            {
                // Da esquerda para a direita --> a coluna varia --> l -> c - 1, a linha permanece constante
                for (i = l; i < c; i++)
                {
                    Console.Write(matrix[k][i] + " ");
                }
                k++;

                // De cima para baixo --> a linha varia --> k -> r - 1, a coluna permanece constante
                for (i = k; i < r; i++)
                {
                    Console.Write(matrix[i][c - 1] + " ");
                }
                c--;

                if (k < r)
                {
                    // Da direita para a esquerda --> a coluna varia --> c - 1 -> l, a linha permanece constante
                    for (i = c - 1; i >= l; i--)
                    {
                        Console.Write(matrix[r - 1][i] + " ");
                    }
                    r--;
                }

                if (l < c)
                {
                    // De baixo para cima --> a linha varia --> r - 1 -> k, a coluna permanece constante
                    for (i = r - 1; i >= k; i--)
                    {
                        Console.Write(matrix[i][l] + " ");
                    }
                    l++;
                }
            }
        }
    }

File: go/dsa/matrix/SpiralMatrix.go
package matrix

import "fmt"

// SpiralPrint imprime a matriz dada em ordem espiral
// matrix: matriz 2D de inteiros
// r: número de linhas na matriz
// c: número de colunas na matriz
func SpiralPrint(matrix [][]int, r int, c int) {

	// Variáveis auxiliares para controlar os limites da matriz
	i := 0  // índice para iteração
	k := 0  // início da linha
	l := 0  // início da coluna

	// Enquanto ainda houver elementos para processar nas submatrizes
	for k < r && l < c {

		// Imprime a primeira linha da submatriz restante
		for i = l; i < c; i++ {
			fmt.Printf("%v"+" ", matrix[k][i])
		}
		k++  // Move o limite superior para a próxima linha

		// Imprime a última coluna da submatriz restante
		for i = k; i < r; i++ {
			fmt.Printf("%v"+" ", matrix[i][c-1])
		}
		c--  // Move o limite direito para a esquerda

		// Verifica se ainda há linhas restantes após o movimento superior
		if k < r {
			// Imprime a última linha da submatriz restante, da direita para a esquerda
			for i := c - 1; i >= l; i-- {
				fmt.Printf("%v"+" ", matrix[r-1][i])
			}
			r--  // Move o limite inferior para cima
		}

		// Verifica se ainda há colunas restantes após o movimento direito
		if l < c {
			// Imprime a primeira coluna da submatriz restante, de baixo para cima
			for i := r - 1; i >= k; i-- {
				fmt.Printf("%v"+" ", matrix[i][l])
			}
			l++  // Move o limite esquerdo para a direita
		}
	}
}
